File: populate-catalog.js
#!/usr/bin/env node
/**
 * populate-catalog.js
 *
 * Fetches appliance products from the Best Buy API and upserts them into
 * the Supabase product_catalog table.
 *
 * Requirements: Node 18+ (uses built-in fetch)
 *
 * Usage:
 *   BESTBUY_API_KEY=xxx \
 *   SUPABASE_URL=https://xxxx.supabase.co \
 *   SUPABASE_SERVICE_KEY=xxx \
 *   node populate-catalog.js
 *
 * Get your Best Buy API key (free): https://developer.bestbuy.com
 * Get your Supabase service key: Supabase dashboard → Project Settings → API → service_role secret
 */

const BESTBUY_API_KEY  = process.env.BESTBUY_API_KEY;
const SUPABASE_URL     = process.env.SUPABASE_URL;
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_KEY;

if (!BESTBUY_API_KEY || !SUPABASE_URL || !SUPABASE_SERVICE_KEY) {
  console.error('Missing required environment variables.');
  console.error('  BESTBUY_API_KEY   — from developer.bestbuy.com');
  console.error('  SUPABASE_URL      — e.g. https://xxxx.supabase.co');
  console.error('  SUPABASE_SERVICE_KEY — from Supabase dashboard → Project Settings → API');
  process.exit(1);
}

// Appliance sub-categories to pull from Best Buy.
// Each entry is a Best Buy category ID that maps to a specific appliance type.
const APPLIANCE_CATEGORIES = [
  { id: 'abcat0901011', label: 'Refrigerators'       },
  { id: 'abcat0901021', label: 'Freezers'             },
  { id: 'abcat0904011', label: 'Washing Machines'     },
  { id: 'abcat0904021', label: 'Dryers'               },
  { id: 'abcat0905011', label: 'Ranges & Ovens'       },
  { id: 'abcat0905021', label: 'Microwaves'           },
  { id: 'abcat0912011', label: 'Dishwashers'          },
  { id: 'abcat0903011', label: 'Air Conditioners'     },
  { id: 'abcat0902011', label: 'Vacuums'              },
  { id: 'abcat0912031', label: 'Trash Compactors'     },
  { id: 'abcat0906011', label: 'Garbage Disposals'    },
];

const PAGE_SIZE  = 100;   // Max allowed by Best Buy API
const BATCH_SIZE = 50;    // Records per Supabase upsert call
const DELAY_MS   = 250;   // Pause between Best Buy API pages to stay polite

// Mirrors the iOS app's ModelNumberNormalizer:
// uppercase, strip spaces, dashes, slashes, dots
function normalizeModelNumber(raw) {
  return raw.toUpperCase().replace(/[\s\-\/\\.]/g, '');
}

async function fetchBestBuyPage(categoryId, page) {
  const filter = `(categoryPath.id=${categoryId}&modelNumber=*&regularPrice>0)`;
  const show   = 'modelNumber,name,regularPrice,manufacturer';
  const url    = `https://api.bestbuy.com/v1/products${filter}?format=json&show=${show}&pageSize=${PAGE_SIZE}&page=${page}&apiKey=${BESTBUY_API_KEY}`;

  const res = await fetch(url);
  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Best Buy API ${res.status}: ${body}`);
  }
  return res.json();
}

async function upsertToSupabase(records) {
  const res = await fetch(`${SUPABASE_URL}/rest/v1/product_catalog`, {
    method: 'POST',
    headers: {
      'Content-Type':  'application/json',
      'apikey':        SUPABASE_SERVICE_KEY,
      'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`,
      'Prefer':        'resolution=merge-duplicates',
    },
    body: JSON.stringify(records),
  });

  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Supabase upsert ${res.status}: ${body}`);
  }
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function processCategory(category) {
  console.log(`\n── ${category.label} (${category.id})`);

  // Fetch page 1 to learn total count
  const first      = await fetchBestBuyPage(category.id, 1);
  const totalPages = first.totalPages ?? 1;
  const total      = first.total ?? 0;

  if (total === 0) {
    console.log('   No products found, skipping.');
    return { inserted: 0, skipped: 0 };
  }

  console.log(`   ${total} products across ${totalPages} pages`);

  let inserted = 0;
  let skipped  = 0;
  let batch    = [];

  const processPage = async (products) => {
    for (const p of products) {
      if (!p.modelNumber || !p.name || !p.regularPrice) {
        skipped++;
        continue;
      }
      const normalized = normalizeModelNumber(p.modelNumber);
      if (!normalized || normalized.length < 3) {
        skipped++;
        continue;
      }
      batch.push({
        normalized_model_number: normalized,
        product_name:            p.name,
        msrp:                    p.regularPrice,
        source:                  'bestbuy-catalog',
        confidence:              1.0,
      });

      if (batch.length >= BATCH_SIZE) {
        await upsertToSupabase(batch);
        inserted += batch.length;
        batch = [];
      }
    }
  };

  await processPage(first.products ?? []);

  for (let page = 2; page <= totalPages; page++) {
    await sleep(DELAY_MS);
    const data = await fetchBestBuyPage(category.id, page);
    await processPage(data.products ?? []);

    if (page % 5 === 0 || page === totalPages) {
      process.stdout.write(`   Page ${page}/${totalPages} — ${inserted} inserted so far\r`);
    }
  }

  // Flush remaining batch
  if (batch.length > 0) {
    await upsertToSupabase(batch);
    inserted += batch.length;
    batch = [];
  }

  console.log(`   ✓ ${inserted} inserted, ${skipped} skipped`);
  return { inserted, skipped };
}

async function main() {
  console.log('ApplianceManifest — Catalog Seeder');
  console.log('====================================');
  console.log(`Supabase: ${SUPABASE_URL}`);
  console.log(`Categories: ${APPLIANCE_CATEGORIES.length}`);

  let totalInserted = 0;
  let totalSkipped  = 0;

  for (const category of APPLIANCE_CATEGORIES) {
    try {
      const { inserted, skipped } = await processCategory(category);
      totalInserted += inserted;
      totalSkipped  += skipped;
    } catch (err) {
      console.error(`\n   ERROR in ${category.label}: ${err.message}`);
    }
  }

  console.log('\n====================================');
  console.log(`Done! Inserted/updated: ${totalInserted} | Skipped: ${totalSkipped}`);
}

main().catch(err => {
  console.error('\nFatal error:', err.message);
  process.exit(1);
});

File: populate-catalog.py
#!/usr/bin/env python3
"""
ApplianceManifest — Catalog Seeder
Fetches appliance products from Best Buy API and upserts into Supabase product_catalog.
"""

import json
import os
import re
import time
import urllib.request
import urllib.parse
import urllib.error

BESTBUY_API_KEY = os.environ.get("BESTBUY_API_KEY")
SUPABASE_URL = os.environ.get("SUPABASE_URL")
SUPABASE_SERVICE_KEY = os.environ.get("SUPABASE_SERVICE_KEY")

# Search by name keyword — proven to work with this API key
APPLIANCE_SEARCHES = [
    ("refrigerator*",  "Refrigerators"),
    ("washer*",        "Washers"),
    ("dryer*",         "Dryers"),
    ("dishwasher*",    "Dishwashers"),
    ("microwave*",     "Microwaves"),
    ("range*",         "Ranges"),
    ("oven*",          "Ovens"),
    ("freezer*",       "Freezers"),
    ("cooktop*",       "Cooktops"),
]

PAGE_SIZE   = 100
BATCH_SIZE  = 50
PAGE_DELAY  = 1.5
RETRY_WAIT  = 65
MAX_RETRIES = 3

def require_env():
    missing = [
        name for name, value in [
            ("BESTBUY_API_KEY", BESTBUY_API_KEY),
            ("SUPABASE_URL", SUPABASE_URL),
            ("SUPABASE_SERVICE_KEY", SUPABASE_SERVICE_KEY),
        ]
        if not value
    ]
    if missing:
        joined = ", ".join(missing)
        raise SystemExit(
            f"Missing required environment variables: {joined}\n"
            "Usage:\n"
            "  BESTBUY_API_KEY=... SUPABASE_URL=https://xxxx.supabase.co "
            "SUPABASE_SERVICE_KEY=... python3 populate-catalog.py"
        )

def normalize_model(raw):
    return re.sub(r'[^A-Z0-9]', '', raw.upper())

def fetch_page(name_keyword, page, attempt=1):
    encoded = urllib.parse.quote(name_keyword)
    url = (
        f'https://api.bestbuy.com/v1/products'
        f'(modelNumber=*&regularPrice>0&name={encoded})'
        f'?format=json'
        f'&show=modelNumber,name,regularPrice'
        f'&pageSize={PAGE_SIZE}'
        f'&page={page}'
        f'&apiKey={BESTBUY_API_KEY}'
    )
    req = urllib.request.Request(url, headers={'User-Agent': 'ApplianceManifest/1.0'})
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        if e.code in (403, 429) and attempt <= MAX_RETRIES:
            print(f'   Rate limited (attempt {attempt}), waiting {RETRY_WAIT}s...', flush=True)
            time.sleep(RETRY_WAIT)
            return fetch_page(name_keyword, page, attempt + 1)
        raise

def upsert_to_supabase(records):
    data = json.dumps(records).encode('utf-8')
    req  = urllib.request.Request(
        f'{SUPABASE_URL}/rest/v1/product_catalog',
        data=data, method='POST',
        headers={
            'Content-Type':  'application/json',
            'apikey':        SUPABASE_SERVICE_KEY,
            'Authorization': f'Bearer {SUPABASE_SERVICE_KEY}',
            'Prefer':        'resolution=merge-duplicates',
        }
    )
    try:
        with urllib.request.urlopen(req, timeout=30):
            pass
    except urllib.error.HTTPError as e:
        raise Exception(f'Supabase {e.code}: {e.read().decode()}')

def process(keyword, label):
    print(f'\n── {label}')
    try:
        first = fetch_page(keyword, 1)
    except Exception as e:
        print(f'   ERROR: {e}')
        return 0, 0

    total       = first.get('total', 0)
    total_pages = first.get('totalPages', 1)

    if total == 0:
        print('   No products found.')
        return 0, 0

    print(f'   {total} products across {total_pages} pages')

    inserted = 0
    skipped  = 0
    batch    = []

    seen_in_run = set()

    def flush():
        nonlocal inserted, batch
        if not batch:
            return
        # Deduplicate within the batch — keep last occurrence (most recent data)
        deduped = {r['normalized_model_number']: r for r in batch}
        # Also skip models already inserted earlier in this run
        unique = [r for k, r in deduped.items() if k not in seen_in_run]
        seen_in_run.update(deduped.keys())
        batch = []
        if unique:
            upsert_to_supabase(unique)
            inserted += len(unique)

    def consume(products):
        nonlocal skipped
        for p in products:
            model = (p.get('modelNumber') or '').strip()
            name  = (p.get('name') or '').strip()
            price = p.get('regularPrice')
            if not model or not name or not price:
                skipped += 1
                continue
            normalized = normalize_model(model)
            if len(normalized) < 3:
                skipped += 1
                continue
            batch.append({
                'normalized_model_number': normalized,
                'product_name':            name,
                'msrp':                    price,
                'source':                  'bestbuy-catalog',
                'confidence':              1.0,
            })
            if len(batch) >= BATCH_SIZE:
                flush()

    consume(first.get('products', []))

    for page in range(2, total_pages + 1):
        time.sleep(PAGE_DELAY)
        try:
            data = fetch_page(keyword, page)
            consume(data.get('products', []))
        except Exception as e:
            print(f'   WARNING page {page}: {e}')
            continue
        if page % 5 == 0 or page == total_pages:
            print(f'   Page {page}/{total_pages} — {inserted + len(batch)} processed', flush=True)

    try:
        flush()
    except Exception as e:
        print(f'   WARNING final flush: {e}')
    print(f'   ✓ {inserted} inserted, {skipped} skipped')
    return inserted, skipped

def main():
    require_env()
    print('ApplianceManifest — Catalog Seeder')
    print('====================================')

    total_in = 0
    total_sk = 0

    for keyword, label in APPLIANCE_SEARCHES:
        ins, skp = process(keyword, label)
        total_in += ins
        total_sk += skp
        time.sleep(PAGE_DELAY)

    print('\n====================================')
    print(f'Done!  Inserted/updated: {total_in}  |  Skipped: {total_sk}')

if __name__ == '__main__':
    main()
